analyzers/src/SonarAnalyzer.CSharp/Rules/VariableShadowsField.cs

```diff
@@ -28,75 +28,130 @@ public sealed class VariableShadowsField : SonarDiagnosticAnalyzer
+        private static List<SyntaxToken> ExtractIdentifiers(SyntaxNode node) =>
+            node switch
+            {
+                VariableDeclarationSyntax variableDeclaration => variableDeclaration.Variables.Select(x => x.Identifier).ToList(),
+                ForEachStatementSyntax foreachStatement => new() { foreachStatement.Identifier },
+                _ when VariableDesignationSyntaxWrapper.IsInstance(node) => ((VariableDesignationSyntaxWrapper)node).AllVariables().Select(x => x.Identifier).ToList(),
+                _ => new()
+            };
+
+        private static List<ISymbol> GetContextSymbols(SonarSyntaxNodeReportingContext context)
         {
-            if (variableDesignation.AllVariables() is { Length: > 0 } variables
-                && context.ContainingSymbol.ContainingType is { } containingType
-                && GetMembers(containingType) is var members)
+            var symbols = context.ContainingSymbol.ContainingSymbol.GetSymbolType().GetMembers();
+            var primaryCtor = symbols.FirstOrDefault(x => x is IMethodSymbol && IsPrimaryCtor(x));
+            var primaryCtorParameters = primaryCtor?.GetParameters();
+            var relevantSymbols = new List<ISymbol>();
+            foreach (var relevant in symbols.Where(x => x is IPropertySymbol or IFieldSymbol))
             {
-                foreach (var variable in variables)
+                if (primaryCtorParameters is not null
+                    && primaryCtorParameters.Any()
+                    && TryGetAssignedValue(relevant, out var assignedValue)
+                    && primaryCtorParameters.Any(x => x.Name == assignedValue))
+                {
+                    primaryCtorParameters = primaryCtorParameters.Where(x => x.Name != assignedValue).ToImmutableArray();
+                }
+                else
                 {
-                    ReportOnVariableMatchingField(context, members, variable.Identifier);
+                    relevantSymbols.Add(relevant);
                 }
             }
+            return primaryCtor is null ? relevantSymbols : relevantSymbols.Concat(primaryCtorParameters).ToList();
         }
 
-        private static void ProcessVariableDeclaration(SonarSyntaxNodeReportingContext context, VariableDeclarationSyntax variableDeclaration)
+        private static bool TryGetAssignedValue(ISymbol symbol, out string assignedValue)
         {
-            if (variableDeclaration is { Variables: { Count: > 0 } variables }
-                && context.ContainingSymbol.ContainingType is { } containingType
-                && GetMembers(containingType) is var members)
+            assignedValue = string.Empty;
+            if (symbol.DeclaringSyntaxReferences.FirstOrDefault()?.GetSyntax() is { } syntax)
             {
-                foreach (var variable in variables)
+                if (syntax is VariableDeclaratorSyntax variableDeclarator && variableDeclarator.Initializer is { Value: IdentifierNameSyntax { } fieldValue })
```

Comment:
This doesn't work when the primary constructor parameter is accessed within en expression like e.g
```cs
object o = o ?? throw new ArgumentException("");
```
You should use `variableDeclarator.Initializer.DescendantNodesAndSelf().OfType<IdentifierNameSyntax>().FirstOrDefault()` (but even that is not correct).

Reply:
I removed this part, see the [next comment](https://github.com/SonarSource/sonar-dotnet/pull/8251#discussion_r1371474909)